File: templates/Py/kmp.py
def build_lps(pa):
  N = len(pa)
  pi = [0] * N
  prev_i = 0
  i = 1
  while i < N:
    if pa[i] == pa[prev_i]:
      pi[i] = prev_i + 1
      i += 1
      prev_i += 1
    else:
      if prev_i == 0:
        pi[i] = 0
        i += 1
      else:
        prev_i = pi[prev_i - 1]
  return pi


if __name__ == "__main__":
  pattern = "abba"
  s = "ababba"
  lps = build_lps(pattern)
  N = len(s)
  i, j, cnt = 0, 0, 0
  while i < N:
    if s[i] == pattern[j]:
      i += 1
      j += 1
    else:
      if j == 0:
        i += 1
      else:
        j = lps[j - 1]
    if j == len(pattern):
      cnt += 1
      j = lps[j - 1]
  assert cnt == 1

File: templates/C++/kmp.cpp
#include <string>
#include <vector>
#include <cassert>
auto build_lps(std::string& s) -> std::vector<int> {
	auto N = s.size();
	auto lps = std::vector<int>(N, 0);
	auto j = 0, i = 1;
	while (i < N) {
		if (s[i] == s[j]) {
			j++;
			lps[i] = j;
			i++;
		} else {
			if (j != 0) {
				j = lps[j - 1];
			} else {
				lps[i] = 0;
				i++;				
			}
		}
	}
	return lps;
}

auto kmp(std::string& s, std::string& p) -> int {
	auto N = s.size();
	auto M = p.size();
	auto lps = build_lps(p);
	auto j = 0, i = 0;
	auto cnt = 0;
	while (i < N) {
		if (s[i] == p[j]) {
			i++;
			j++;
		} else {
			if (j != 0) {
				j = lps[j - 1];
			} else {
				i++;
			}
		}
		if (j == M) {
			cnt++;
			j = lps[j - 1];
		}
	}
	return cnt;
}
int main(void) {
    auto s = std::string("abcabcalsdkfj;lab");
    auto p = std::string("ab");
    auto ret = kmp(s, p);
    assert(ret == 3);
}